File: 67. Mobile MDM-lite device compliance checker app/mobile/android-agent/java/com/mdmlite/agent/ComplianceEngine.java
package com.mdmlite.agent;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

/**
 * Compliance evaluation, mirroring the desktop engine: severity-weighted score,
 * critical-fail gate, and a final COMPLIANT / NON_COMPLIANT verdict.
 */
public final class ComplianceEngine {

    private static int weight(String severity) {
        switch (severity == null ? "" : severity) {
            case "low":      return 1;
            case "medium":   return 2;
            case "high":     return 3;
            case "critical": return 4;
            default:         return 1;
        }
    }

    public static JSONObject evaluate(JSONObject telemetry) throws JSONException {
        List<BaselinePolicy.Rule> rules = BaselinePolicy.RULES;
        JSONArray results = new JSONArray();
        int totalW = 0, passW = 0, pass = 0, fail = 0, na = 0;
        boolean criticalFail = false;

        for (BaselinePolicy.Rule rule : rules) {
            boolean applicable = !rule.id.startsWith("ios");
            String verdict = "NA";
            String actual = "-";
            int w = weight(rule.severity);

            if (applicable) {
                switch (rule.kind) {
                    case "version_gte": {
                        String os = telemetry.optJSONObject("os").optString("version", "0.0");
                        actual = os;
                        verdict = atLeast(os, String.valueOf(rule.expected)) ? "PASS" : "FAIL";
                        break;
                    }
                    case "boolean": {
                        JSONObject sec = telemetry.optJSONObject("security");
                        String key = secKey(rule.id);
                        boolean val = sec != null && sec.optBoolean(key, false);
                        actual = String.valueOf(val);
                        boolean expected = Boolean.parseBoolean(String.valueOf(rule.expected));
                        verdict = val == expected ? "PASS" : "FAIL";
                        break;
                    }
                    case "allowlist": {
                        JSONObject apps = telemetry.optJSONObject("apps");
                        JSONArray installed = apps == null ? new JSONArray() : apps.optJSONArray("installed");
                        String[] required = (String[]) rule.expected;
                        JSONArray missing = new JSONArray();
                        for (String req : required) {
                            if (installed == null || !contains(installed, req)) missing.put(req);
                        }
                        actual = missing.length() == 0 ? "-" : missing.toString();
                        verdict = missing.length() == 0 ? "PASS" : "FAIL";
                        break;
                    }
                    case "denylist": {
                        JSONObject apps = telemetry.optJSONObject("apps");
                        JSONArray installed = apps == null ? new JSONArray() : apps.optJSONArray("installed");
                        String[] banned = (String[]) rule.expected;
                        JSONArray found = new JSONArray();
                        for (String b : banned) {
                            if (installed != null && contains(installed, b)) found.put(b);
                        }
                        actual = found.length() == 0 ? "-" : found.toString();
                        verdict = found.length() == 0 ? "PASS" : "FAIL";
                        break;
                    }
                    default:
                        verdict = "NA";
                }

                if ("PASS".equals(verdict)) { totalW += w; passW += w; pass++; }
                else if ("FAIL".equals(verdict)) { totalW += w; fail++;
                    if ("critical".equals(rule.severity)) criticalFail = true; }
                else na++;
            } else {
                na++;
            }

            results.put(new JSONObject()
                    .put("rule_id", rule.id)
                    .put("group", rule.group)
                    .put("label", rule.label)
                    .put("severity", rule.severity)
                    .put("verdict", verdict)
                    .put("expected", rule.expected)
                    .put("actual", actual));
        }

        double ratio = totalW == 0 ? 0.0 : (double) passW / totalW;
        boolean compliant = !criticalFail && ratio >= BaselinePolicy.THRESHOLD;

        return new JSONObject()
                .put("status", compliant ? "COMPLIANT" : "NON_COMPLIANT")
                .put("score", Math.round(ratio * 1000.0) / 10.0)
                .put("pass_count", pass)
                .put("fail_count", fail)
                .put("na_count", na)
                .put("threshold", BaselinePolicy.THRESHOLD * 100)
                .put("results", results);
    }

    private static String secKey(String ruleId) {
        switch (ruleId) {
            case "android.security.root":          return "rooted";
            case "android.security.encryption":    return "encryption";
            case "android.security.screenlock":    return "screen_lock";
            case "android.security.unknownsources":return "unknown_sources";
            case "android.security.playprotect":   return "play_protect";
            default:                               return "screen_lock";
        }
    }

    private static boolean contains(JSONArray arr, String needle) {
        for (int i = 0; i < arr.length(); i++) {
            if (needle.equals(arr.optString(i))) return true;
        }
        return false;
    }

    private static boolean atLeast(String actual, String minimum) {
        return compareVersion(actual, minimum) >= 0;
    }

    private static int compareVersion(String a, String b) {
        long[] va = versionParts(a);
        long[] vb = versionParts(b);
        for (int i = 0; i < 3; i++) {
            if (va[i] != vb[i]) return Long.compare(va[i], vb[i]);
        }
        return 0;
    }

    private static long[] versionParts(String v) {
        long[] out = new long[3];
        if (v == null) return out;
        String[] parts = v.replaceAll("[^0-9.]", "").split("\\.");
        for (int i = 0; i < parts.length && i < 3; i++) {
            try { out[i] = Long.parseLong(parts[i]); } catch (NumberFormatException ignore) { }
        }
        return out;
    }
}

File: 67. Mobile MDM-lite device compliance checker app/mobile/ios/MDM-LiteAgent/Sources/MDM-LiteAgent/ComplianceEngine.swift
import Foundation
import LocalAuthentication

/// iOS pipeline: collect telemetry -> evaluate against the baseline policy ->
/// produce a COMPLIANT / NON_COMPLIANT verdict (mirror of the desktop engine).
public struct ComplianceEngine {

    public struct Rule {
        public let id: String
        public let group: String
        public let label: String
        public let severity: String
        public let kind: String
        public let expected: Any
    }

    public static let threshold = 0.80

    // Mirror of "MDM-Lite Baseline Policy" (iOS subset + shared checks).
    public static let baseline: [Rule] = [
        Rule(id: "ios.os.version", group: "os", label: "iOS version",
             severity: "high", kind: "version_gte", expected: "15.0"),
        Rule(id: "ios.security.jailbreak", group: "security",
             label: "Jailbreak detection",
             severity: "critical", kind: "boolean", expected: false),
        Rule(id: "ios.security.encryption", group: "security",
             label: "Data protection",
             severity: "high", kind: "boolean", expected: true),
        Rule(id: "ios.security.passcode", group: "security",
             label: "Passcode enforced",
             severity: "high", kind: "boolean", expected: true),
        Rule(id: "ios.security.allowUntrusted", group: "security",
             label: "Untrusted certificates blocked",
             severity: "medium", kind: "boolean", expected: false),
    ]

    /// Returns a dict with status / score / counts / per-rule results.
    public func evaluate(telemetry: [String: Any]) -> [String: Any] {
        var results: [[String: Any]] = []
        var totalW = 0, passW = 0, pass = 0, fail = 0
        var criticalFail = false

        let os = (telemetry["os"] as? [String: Any]) ?? [:]
        let sec = (telemetry["security"] as? [String: Any]) ?? [:]

        for rule in Self.baseline {
            let verdict = evaluate(rule: rule, os: os, security: sec)
            let weight = Self.weight(of: rule.severity)
            switch verdict {
            case "PASS":
                totalW += weight; passW += weight; pass += 1
            case "FAIL":
                totalW += weight; fail += 1
                if rule.severity == "critical" { criticalFail = true }
            default:
                break
            }
            results.append([
                "rule_id": rule.id, "group": rule.group, "label": rule.label,
                "severity": rule.severity, "verdict": verdict,
                "expected": rule.expected, "actual": actualValue(rule, os: os, security: sec),
            ])
        }

        let ratio = totalW == 0 ? 0.0 : Double(passW) / Double(totalW)
        let compliant = !criticalFail && ratio >= Self.threshold
        return [
            "status": compliant ? "COMPLIANT" : "NON_COMPLIANT",
            "score": Int((ratio * 1000).rounded()) / 10,
            "pass_count": pass, "fail_count": fail,
            "na_count": Self.baseline.count - pass - fail,
            "threshold": Int(Self.threshold * 100),
            "results": results,
        ]
    }

    private func evaluate(rule: Rule, os: [String: Any], security: [String: Any]) -> String {
        switch rule.kind {
        case "version_gte":
            let actual = (os["version"] as? String) ?? "0.0"
            return version(actual) >= version(rule.expected as? String ?? "0.0") ? "PASS" : "FAIL"
        case "boolean":
            guard let key = securityKey(for: rule.id) else { return "NA" }
            let actual = (security[key] as? Bool) ?? false
            let expected = (rule.expected as? Bool) ?? false
            return actual == expected ? "PASS" : "FAIL"
        default:
            return "NA"
        }
    }

    private func actualValue(_ rule: Rule, os: [String: Any], security: [String: Any]) -> Any? {
        switch rule.kind {
        case "version_gte": return os["version"]
        case "boolean":
            guard let key = securityKey(for: rule.id) else { return nil }
            return security[key]
        default: return nil
        }
    }

    private func securityKey(for ruleId: String) -> String? {
        switch ruleId {
        case "ios.security.jailbreak": return "jailbreak"
        case "ios.security.encryption": return "encryption"
        case "ios.security.passcode": return "screen_lock"
        case "ios.security.allowUntrusted": return "side_loading"
        default: return nil
        }
    }

    private func version(_ v: String) -> [Int] {
        let digits = v.filter { $0.isNumber || $0 == "." }
        var out = digits.split(separator: ".").prefix(3).map { Int($0) ?? 0 }
        while out.count < 3 { out.append(0) }
        return out
    }

    private static func weight(of severity: String) -> Int {
        switch severity {
        case "low": return 1
        case "medium": return 2
        case "high": return 3
        case "critical": return 4
        default: return 1
        }
    }
}